File: code/richediting/src/main/java/org/fbme/ide/richediting/adapters/fbnetwork/FBNetworkComponentController.java
package org.fbme.ide.richediting.adapters.fbnetwork;

import org.fbme.ide.richediting.viewmodel.NetworkPortView;
import org.jetbrains.annotations.NotNull;

import java.awt.*;

public interface FBNetworkComponentController {

    @NotNull
    Rectangle getBounds(@NotNull Point position);

    @NotNull
    default Rectangle getFBCellBounds(@NotNull Point position) {
        return getBounds(position);
    };

    @NotNull
    Point getPortCoordinates(@NotNull NetworkPortView port, @NotNull Point position);

    @NotNull
    Rectangle getPortBounds(@NotNull NetworkPortView port, @NotNull Point position);

    boolean isSource(@NotNull NetworkPortView port);

    default boolean canBeSourcedAt(@NotNull NetworkPortView port, @NotNull Point position) {
        return isSource(port);
    }

    default boolean canBeTargetedAt(@NotNull NetworkPortView port, @NotNull Point position) {
        return !isSource(port);
    }
}

File: code/richediting/src/main/kotlin/org/fbme/ide/richediting/adapters/fbnetwork/FBNetworkComponentController.kt
package org.fbme.ide.richediting.adapters.fbnetwork

import org.fbme.ide.richediting.viewmodel.NetworkPortView
import java.awt.Point
import java.awt.Rectangle

interface FBNetworkComponentController {
    fun getBounds(position: Point): Rectangle
    fun getFBCellBounds(position: Point): Rectangle {
        return getBounds(position)
    }

    fun getFBPortTemplates(): Set<NetworkPortView> {
        return emptySet()
    }

    fun getPortCoordinates(port: NetworkPortView, position: Point): Point
    fun getPortBounds(port: NetworkPortView, position: Point): Rectangle
    fun isSource(port: NetworkPortView): Boolean
    fun canBeSourcedAt(port: NetworkPortView, position: Point): Boolean {
        return isSource(port)
    }

    fun canBeTargetedAt(port: NetworkPortView, position: Point): Boolean {
        return !isSource(port)
    }

    fun getTemplateBounds(template: NetworkPortView, modelForm: Point): Rectangle {
        return Rectangle(0, 0)
    }

    fun getTemplatePosition(template: NetworkPortView, modelForm: Point): Point {
        return Point(0, 0)
    }

    fun createPort(source: NetworkPortView, template: NetworkPortView): NetworkPortView? {
        return null
    }
}
